svm-rent-collector/src/svm_rent_collector.rs

```diff
@@ -0,0 +1,137 @@
+/// management.
+pub trait SVMRentCollector {
+    /// Check rent state transition for an account in a transaction.
+    ///
+    /// This method has a default implementation that calls into
+    /// `check_rent_state_with_account`.
+    fn check_rent_state(
+        &self,
+        pre_rent_state: Option<&RentState>,
+        post_rent_state: Option<&RentState>,
+        transaction_context: &TransactionContext,
+        index: IndexOfAccount,
+    ) -> Result<()> {
+        if let Some((pre_rent_state, post_rent_state)) = pre_rent_state.zip(post_rent_state) {
+            let expect_msg =
+                "account must exist at TransactionContext index if rent-states are Some";
+            self.check_rent_state_with_account(
+                pre_rent_state,
+                post_rent_state,
+                transaction_context
+                    .get_key_of_account_at_index(index)
+                    .expect(expect_msg),
+                &transaction_context
+                    .get_account_at_index(index)
+                    .expect(expect_msg)
+                    .borrow(),
+                index,
+            )?;
+        }
+        Ok(())
+    }
+
+    /// Check rent state transition for an account directly.
+    ///
+    /// This method has a default implementation that checks whether the
+    /// transition is allowed and returns an error if it is not. It also
+    /// verifies that the account is not the incinerator.
+    fn check_rent_state_with_account(
+        &self,
+        pre_rent_state: &RentState,
+        post_rent_state: &RentState,
+        address: &Pubkey,
+        _account_state: &AccountSharedData,
+        account_index: IndexOfAccount,
+    ) -> Result<()> {
+        if !solana_sdk::incinerator::check_id(address)
+            && !self.transition_allowed(pre_rent_state, post_rent_state)
+        {
+            let account_index = account_index as u8;
+            Err(TransactionError::InsufficientFundsForRent { account_index })
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Collect rent from an account.
+    fn collect_rent(&self, address: &Pubkey, account: &mut AccountSharedData) -> CollectedInfo;
+
+    /// Determine the rent state of an account.
+    ///
+    /// This method has a default implementation that treats accounts with zero
+    /// lamports as uninitialized and uses the implemented `get_rent` to
+    /// determine whether an account is rent-exempt.
+    fn get_account_rent_state(&self, account: &AccountSharedData) -> RentState {
+        if account.lamports() == 0 {
+            RentState::Uninitialized
+        } else if self
+            .get_rent()
+            .is_exempt(account.lamports(), account.data().len())
+        {
+            RentState::RentExempt
+        } else {
+            RentState::RentPaying {
+                data_size: account.data().len(),
+                lamports: account.lamports(),
+            }
+        }
+    }
```

Comment:
I've taken the methods from `RentState` and implemented them on the trait itself to provide more customization.